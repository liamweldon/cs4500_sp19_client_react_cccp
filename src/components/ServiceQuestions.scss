.questions {
  button {
    background: none;
    border: none;
  }
  i {
    font-size: 2em;

    &.fa-trash-alt {
      color: red;
    }
    &.fa-pen-square,
    &.fa-plus-square {
      color: blue;
    }
    &.fa-check-square {
      color: green;
    }
  }
}
