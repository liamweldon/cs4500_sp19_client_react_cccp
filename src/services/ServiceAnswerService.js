export default class ServiceAnswerService {
    static instance = null;
    static getInstance() {
        if(ServiceAnswerService.instance === null) {
            ServiceAnswerService.instance = new ServiceAnswerService()
        }
        return this.instance
    }
    findServiceAnswerById = id =>
    fetch(`http://localhost:8080/api/service-answers/${id}`)
    .then(response => response.json())
    findAllServiceAnswers = () =>
    fetch("http://localhost:8080/api/service-answers")
    .then(response => response.json())
}