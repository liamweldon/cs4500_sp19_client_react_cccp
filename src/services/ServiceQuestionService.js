import {API_ROOT} from '../api-config';

export default class ServiceQuestionService {
    static instance = null;
    static getInstance() {
        if (ServiceQuestionService.instance === null) {
            ServiceQuestionService.instance = new ServiceQuestionService();
        }
        return this.instance;
    }
    findServiceQuestionById = (id) =>
        fetch(`${API_ROOT}/api/service-questions/${id}`).then((response) => response.json());

    findAllServiceQuestions = () => fetch(`${API_ROOT}/api/service-questions`).then((response) => response.json());

    addServiceQuestion = (newServiceQuestion) =>
    fetch(`${API_ROOT}/api/service-questions/`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json'
      },
      body: JSON.stringify(newServiceQuestion)
    }).then((response) => response.json());
}
