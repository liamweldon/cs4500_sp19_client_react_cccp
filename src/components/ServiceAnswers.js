import React from 'react'
import ServiceAnswerService from '../services/ServiceAnswerService'
class ServiceAnswers extends React.Component {
    constructor(props) {
        super(props)
        this.serviceAnswerService = ServiceAnswerService.getInstance()
        this.state = {
            serviceAnswers: []
        }
    }
    componentDidMount() {
        this.serviceAnswerService
            .findAllServiceAnswers()
            .then(serviceAnswers =>
                this.setState({
                    serviceAnswers: serviceAnswers
                })
            )
    }
    render() {
        return(
            <div>
                <h3>Service Answers</h3>
                <table className="table">
                    <tbody>
                    {
                        this.state.serviceAnswers
                            .map(serviceAnswer =>
                                <tr key={serviceAnswer.id}>
                                    <td>{serviceAnswer.choiceAnswer}</td>
                                </tr>
                            )
                    }
                    </tbody>
                </table>
            </div>
        )
    }
}

export default ServiceAnswers
