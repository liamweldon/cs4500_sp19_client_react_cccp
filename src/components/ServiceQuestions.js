import React from 'react'
import ServiceQuestionService from '../services/ServiceQuestionService'
import './table.scss';
import {Link} from 'react-router-dom';
class ServiceQuestions extends React.Component {
    constructor(props) {
        super(props)
        this.serviceQuestionService = ServiceQuestionService.getInstance()
        this.state = {
            serviceQuestions: [],
            newServiceQuestion: {
              serviceType: '',
              type: '',
              question: ''
            }
        }
    }

    getServiceQuestions = () => {
      this.serviceQuestionService.findAllServiceQuestions().then((questions) =>
        this.setState({
          serviceQuestions: questions
        })
      );
    };

    componentDidMount() {
        this.serviceQuestionService
            .findAllServiceQuestions()
            .then(serviceQuestions =>
                this.setState({
                    serviceQuestions: serviceQuestions
                })
            )
    }

    handleNewServiceQuestionInputChange = (e) => {
      this.setState({newServiceQuestion: {...this.state.newServiceQuestion, [e.target.name]: e.target.value}});
    };

    addServiceQuestion = () => {
      const newServiceQuestion = this.state.newServiceQuestion;
      this.ServiceQuestionService.addServiceQuestion(newServiceQuestion).then((res) => {
        this.getServiceQuestions();
        this.setState({
          newServiceQuestion: {
          serviceType: '',
          type: '',
          question: ''
        }
      });
    });
  };

    render() {
        return(
            <div>
                <h3>Service Questions</h3>
                <table className="table">
                    <tbody>
                    <tr className="header-row">
                            <td>Type</td>
                            <td>Question</td>
                    </tr>
                    <tr key={-1}>
                      <td>
                        <input value={this.state.newServiceQuestion.service} name="service" onChange={this.handleNewServiceQuestionInputChange} />
                      </td>
                      <td>
                        <select
                          value={this.state.newServiceQuestion.type}
                          name="type"
                          onChange={this.handleNewServiceQuestionInputChange}
                        />
                      </td>
                      <td>
                        <input
                          value={this.state.newServiceQuestion.question}
                          name="question"
                          onChange={this.handleNewServiceQuestionInputChange}
                        />
                      </td>


                    </tr>
                    {this.state.serviceQuestions.map((serviceQuestion) => (
                            <tr key={serviceQuestion.id}>
                                <td>
                                    {serviceQuestion.type}
                                </td>
                                <td><Link to={`/admin/service-questions/${serviceQuestion.id}`}>{serviceQuestion.question}</Link></td>
                            </tr>
                        ))}
                    </tbody>
                </table>
            </div>
        )
    }
}

export default ServiceQuestions
