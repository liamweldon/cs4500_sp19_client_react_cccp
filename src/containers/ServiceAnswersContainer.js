import React from 'react';
import ServiceAnswerService from '../services/ServiceAnswerService';
import ServiceQuestionService from '../services/ServiceQuestionService';
import '../components/table.scss';
import '../components/ServiceQuestions.scss';
import {omit} from 'lodash';
import ServiceAnswers from '../components/ServiceAnswers.js';

class ServiceAnswersContainer extends React.Component {
    constructor(props) {
        super(props)
        this.serviceAnswerService = ServiceAnswerService.getInstance()
        this.serviceQuestionService = ServiceQuestionService.getInstance()
        this.state = {
            serviceAnswers: [],
            serviceQuestions: [],
            editing: {},
            newAnswer: {
                 type: '',
                 question: '',
                 answer: ''
               }
        }
    }
    componentDidMount() {
        this.serviceAnswerService
            .findAllServiceAnswers()
            .then(serviceAnswers =>
                this.setState({
                    serviceAnswers: serviceAnswers
                })
            );
        this.serviceQuestionService
            .findAllServiceQuestions()
            .then(serviceQuestions =>
                this.setState({
                    serviceQuestions: serviceQuestions
                })
            );
        }

      getServiceAnswers = () => {
        this.serviceAnswerService.findAllServiceAnswers().then((answers) =>
          this.setState({
            serviceAnswers: answers
          })
        );
      };

    saveServiceAnswer = (id) => {
      const newAnswer = this.state.editing[id];
      this.serviceAnswerService.editServiceAnswer(id, newAnswer).then((res) => {
        this.getServiceAnswers();
        this.setState({editing: omit(this.state.editing, id)});
      });
    };

      handleInputChange = (e, id) => {
        let editingCopy = {...this.state.editing};
        editingCopy[id] = {...editingCopy[id], [e.target.name]: e.target.value};
        this.setState({editing: editingCopy});
      };

      handleNewServiceAnswerInputChange = (e) => {
        this.setState({newAnswer: {...this.state.newAnswer, [e.target.name]: e.target.value}});
      };

      addServiceAnswer = () => {
        const newAnswer = this.state.newAnswer;
        this.serviceAnswerService.addServiceAnswer(newAnswer).then((res) => {
          this.getServiceAnswers();
          this.setState({
            newAnswer: {
              type: '',
              question: '',
              answer: ''
          }
        });
      });
    };

      deleteServiceAnswer = (id) => {
        // once the answer is deleted, reload the answers so the user doesnt see the deleted one
        this.serviceAnswerService.deleteServiceAnswer(id).then((res) => this.getServiceAnswers());
      };

    toggleEditMode = (serviceAnswer) => {
      if (serviceAnswer.id in this.state.editing) {
        this.setState({editing: omit(this.state.editing, serviceAnswer.id)});
      } else {
        this.setState({
          editing: {...this.state.editing, [serviceAnswer.id]: {
            type: serviceAnswer.type,
            question: serviceAnswer.question,
            answer: serviceAnswer.choiceAnswer}
          }
        });
      }
    };

      render() {
          const eventHandlers = {
            handleNewServiceAnswerInputChange: this.handleNewServiceAnswerInputChange,
            handleInputChange: this.handleInputChange,
            addServiceAnswer: this.addServiceAnswer,
            deleteServiceAnswer: this.deleteServiceAnswer,
            toggleEditMode: this.toggleEditMode,
            saveServiceAnswer: this.saveServiceAnswer
          }

        return (
          <ServiceAnswers
            serviceAnswers={this.state.serviceAnswers}
            serviceQuestions={this.state.serviceQuestions}
            newAnswer={this.state.newAnswer}
            editing={this.state.editing}
            eventHandlers={eventHandlers}
            />
        );
      }
    }
export default ServiceAnswersContainer;
