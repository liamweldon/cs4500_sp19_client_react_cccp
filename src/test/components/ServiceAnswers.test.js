import React from 'react'
import renderer from 'react-test-renderer';
import ServiceAnswers from '../../components/ServiceAnswers'
import ServiceAnswersContainer from '../../containers/ServiceAnswersContainer'
import ServiceAnswerService from '../../services/ServiceAnswerService'
import ServiceQuestions from '../../components/ServiceQuestions'
import ServiceQuestionService from '../../services/ServiceQuestionService'
import {StaticRouter} from 'react-router';
import Enzyme, {mount} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import serviceAnswers from '../mockData/serviceanswer.mock.json'
import mockServiceAnswer from '../services/mocks/MockServiceAnswerService'
import serviceQuestions from '../mockData/servicequestion.mock.json'

Enzyme.configure({adapter: new Adapter()});

// Baseline initial service answer
let mock_initial_service_answer = {id: 0, type: "Type", question: "Question", answer: "Answer"}

// Mock service answer services
const serviceAnswerService = ServiceAnswerService.getInstance();
const serviceQuestionService = ServiceQuestionService.getInstance();
beforeEach(() => {
	global.fetch = mockServiceAnswer(serviceAnswers);
})

test('Service Answer Summary Snapshot and DOM', () => {
  const component = renderer.create(
    <StaticRouter location="/admin/service-answers" context={{}}>
			<ServiceAnswers serviceAnswers={serviceAnswers} serviceQuestions={serviceQuestions}
				editing={{}} eventHandlers={{}} newAnswer={mock_initial_service_answer}/>
    </StaticRouter>
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();

	let testInstance = component.root;
	let rows = testInstance.findAllByProps({className : "answer-row"});
	expect(rows.length).toBe(2);
});
