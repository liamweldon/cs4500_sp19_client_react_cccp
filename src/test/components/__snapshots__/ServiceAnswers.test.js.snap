// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Service Answer Summary Snapshot and DOM 1`] = `
<div>
  <h3>
    Service Answers
  </h3>
  <table
    className="table"
  >
    <tbody>
      <tr
        className="header-row"
      >
        <td>
          Question
        </td>
        <td>
          Type
        </td>
        <td>
          Answer
        </td>
      </tr>
      <tr>
        <select
          name="question"
          value="Question"
        >
          <option
            value="Can you take care of both cats and dogs?"
          >
            Can you take care of both cats and dogs?
          </option>
          <option
            value="Do you own a hammer?"
          >
            Do you own a hammer?
          </option>
        </select>
        <td>
          <select
            name="type"
          >
            <option
              value="Choice"
            >
              Choice
            </option>
            <option
              value="True/False"
            >
              True/False
            </option>
            <option
              value="Range"
            >
              Range
            </option>
          </select>
        </td>
        <td>
          <input
            name="answer"
            value="Answer"
          />
        </td>
        <td>
          <button>
            Create +
          </button>
        </td>
      </tr>
      <tr
        className="answer-row"
      >
        <td>
          how many bathrooms are there?
        </td>
        <td>
          Choice
        </td>
        <td>
          <a
            href="/admin/service-answers/1"
            onClick={[Function]}
          >
            1
          </a>
        </td>
        <td>
          <button
            className="delete-btn"
            onClick={[Function]}
          >
            <i
              className="fas fa-trash-alt"
            />
          </button>
        </td>
      </tr>
      <tr
        className="answer-row"
      >
        <td>
          when are you available?
        </td>
        <td>
          Choice
        </td>
        <td>
          <a
            href="/admin/service-answers/2"
            onClick={[Function]}
          >
            2
          </a>
        </td>
        <td>
          <button
            className="delete-btn"
            onClick={[Function]}
          >
            <i
              className="fas fa-trash-alt"
            />
          </button>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`ServiceAnswers Summary Snapshots 1`] = `
            <div>
                <h3>Service Answers</h3>
                <table className="table">
                    <tbody>
                    <tr className="header-row">
                            <td>Question</td>
                            <td>Type</td>
                            <td>Answer</td>
                         </tr>
                        <tr key={-1}>
                        <select
                            value="Question"
                            name="question">
                            <option value="Do you have previous experience?">Do you have previous experience?</option>
                            <option value="Do you live within a 20 mile radius?">Do you live within a 20 mile radius?</option>
                            <option value="Are you available within the next week?">Are you available within the next week?</option>
                            <option value="Do you take cash?">Do you take cash?</option>
                        </select>
                          <td>
                            <select
                              name="type"
                              value="Type">
                              <option value="Choice">Choice</option>
                              <option value="True/False">True/False</option>
                              <option value="Range">Range</option>
                            </select>
                          </td>
                          <td>
                            <input
                              value="Answer"
                              name="answer"
                            />
                          </td>
                        <td>
                          <button onClick={[Function]}>
                            Create +
                          </button>
                        </td>
                        </tr>
                        <tr className="answer-row">
                        <td></td>
                        <td></td>
                        <td>
                        <a href="/admin/service-answers/1">1</a>
                        </td>
                        <td>
                          <button
                            className="delete-btn"
                            onClick={[Function]}
                          >
                            <i className="fas fa-trash-alt" />
                          </button>
                          </td>
                        </tr>
                        <tr className="answer-row">
                        <td></td>
                        <td></td>
                        <td>
                        <a href="/admin/service-answers/2">2</a>
                        </td>
                        <td>
                          <button
                            className="delete-btn"
                            onClick={[Function]}
                          >
                            <i className="fas fa-trash-alt" />
                          </button>
                          </td>
                        </tr>
                    </tbody>
                </table>
            </div>
  `;
